docs/get-started/configuration.md
# Configuration
Create a `comments.php` file under your `/config` directory with the following options available to you. You can also use multi-environment options to change these per environment.

The below shows the defaults already used by Comments, so you don't need to add these options unless you want to modify the values.

```php
<?php

return [
    '*' => [
        'indexSidebarLimit' => 25,
        'indexSidebarGroup' => true,
        'indexSidebarIndividualElements' => false,
        'defaultQueryStatus' => ['approved'],

        // General
        'allowGuest' => false,
        'guestNotice' => '',
        'guestRequireEmailName' => true,
        'guestShowEmailName' => true,
        'requireModeration' => true,
        'moderatorUserGroup' => 'xxxxxxxxxxxxxxxxxxxxx',
        'moderatorExcluded' => true,
        'autoCloseDays' => '',
        'maxReplyDepth' => '',
        'maxUserComments' => '',

        // Voting
        'allowVoting' => true,
        'allowGuestVoting' => false,
        'downvoteCommentLimit' => 5,
        'hideVotingForThreshold' => false,

        // Flagging
        'allowFlagging' => true,
        'allowGuestFlagging' => false,
        'flaggedCommentLimit' => 5,

        // Templates - Default
        'showAvatar' => true,
        'placeholderAvatar' => '',
        'showTimeAgo' => true,
        'outputDefaultCss' => true,
        'outputDefaultJs' => true,

        // Templates - Custom
        'templateFolderOverride' => '',
        'templateEmail' => '',

        // Security
        'enableSpamChecks' => true,
        'securityMaxLength' => '',
        'securityFlooding' => '',
        'securityModeration' => '',
        'securitySpamlist' => '',
        'securityBanned' => '',
        'securityMatchExact' => false,
        'recaptchaEnabled' => false,
        'recaptchaKey' => '',
        'recaptchaSecret' => '',
        'recaptchaMinScore' => 0.5,

        // Notifications
        'notificationAuthorEnabled' => true,
        'notificationReplyEnabled' => true,
        'notificationSubscribeAuto' => false,
        'notificationSubscribeDefault' => true,
        'notificationSubscribeEnabled' => false,
        'notificationSubscribeCommentEnabled' => false,
        'notificationModeratorEnabled' => false,
        'notificationModeratorEditEnabled' => false,
        'notificationModeratorApprovedEnabled' => false,
        'notificationAdmins' => [],
        'notificationAdminEnabled' => false,
        'notificationFlaggedEnabled' => false,
        'useQueueForNotifications' => false,

        // Permissions
        'permissions' => [],

        // Custom Fields
        'showCustomFieldNames' => false,
        'showCustomFieldInstructions' => false,

        // CP Sort
        'sortDefaultKey' => 'structure',
        'sortDefaultDirection' => 'asc',
    ]
];
```

## Configuration options
- `indexSidebarLimit` - Set a limit for the number of elements in the comments index sidebar in the control panel.
- `indexSidebarGroup` - Whether to group elements in the comments index sidebar in the control panel.
- `indexSidebarIndividualElements` - Whether to show individual elements in the comments index sidebar in the control panel.
- `defaultQueryStatus` - Set the default status for element queries to return.

- `allowGuest` - Whether to allow guest commenting.
- `guestNotice` - When guest commenting is not allowed, and the user is a guest, display this message in place of the form.
- `guestRequireEmailName` - Whether guests should be required to enter their name and email.
- `guestShowEmailName` - Whether guests should be shown fields to enter their name and email.
- `requireModeration` - Whether comments should be moderated before being public.
- `moderatorUserGroup` - The UID of the User Group that should moderate comments and receive notifications.
- `moderatorExcluded` - Comments from any user in the moderator group will not require moderation themselves.
- `autoCloseDays` - Number of days until commenting is automatically closed. This uses the "Post Date" of an element. 0 to disable.
- `maxReplyDepth` - Set the number of levels (depth) replies for comments can have. Leave empty for no restrictions, 0 to disable replies, or any number to limit how many levels of replies can be made.
- `maxUserComments` - Set the number of comments each user is allowed for each owner element. Leave empty for no restrictions.

- `allowVoting` - Whether to allow voting.
- `allowGuestVoting` - Whether to allow guest voting.
- `downvoteCommentLimit` - Number of down votes required for comment to be marked as `isPoorlyRated`.
- `hideVotingForThreshold` - Whether to hide voting altogether when `isPoorlyRated` is true.

- `allowFlagging` - Whether to allow flagging.
- `allowGuestFlagging` - Whether to allow guest flagging.
- `flaggedCommentLimit` - Number of flags required for comment to be marked as `isFlagged`.

- `showAvatar` - Whether to show an avatar for comments
- `placeholderAvatar` - When "Show avatar" is enabled, and a guest is making a comment, show this as their avatar.
- `showTimeAgo` - Whether to show how long ago a comment was made.
- `outputDefaultCss` - Whether to output the default CSS for the form.
- `outputDefaultJs` - Whether to output the default JS for the form.

- `templateFolderOverride` - Provide a path to your own templates in the below folder.
- `templateEmail` - The template Comments will use for HTML emails.

- `enableSpamChecks` - Whether to enable spam checks for comments.
- `securityMaxLength` - The maximum number of characters a single comment can have. 
- `securityFlooding` - The number of seconds between commenting.
- `securityModeration` - A collection of words that if entered require comments to be moderated.
- `securitySpamlist` - A collection of words that if entered mark comments as spam.
- `securityBanned` - A collection of words that if entered mark comments as trashed.
- `securityMatchExact` - Whether to enable exact keyword matching. With this turned on, it will no longer match words within other words (e.g. ‘craft’ will not match ‘crafty’).
- `recaptchaKey` - The required key for reCAPTCHA.
- `recaptchaSecret` - The required secret for reCAPTCHA.
- `recaptchaMinScore` - The minimum "score" value returned from reCAPTCHA to rate the comment as spam.

- `notificationAuthorEnabled` - Whether to notify element authors when a comment is made.
- `notificationReplyEnabled` - Whether to notify comment authors when a reply is made.
- `notificationSubscribeAuto` - Whether to automatically subscribe to notifications on any comments on the same element, after your first reply.
- `notificationSubscribeDefault` - Whether to automatically subscribe to notifications on comments that the user owns.
- `notificationSubscribeEnabled` - Whether to allow subscriber notification altogether.
- `notificationSubscribeCommentEnabled` - Users can subscribe to a specific thread of comments made on an element.
- `notificationModeratorEnabled` - Whether a nominated user group should receive an email when a new comment has been made, and needs to be moderated.
- `notificationModeratorEditEnabled` - Whether moderators should receive a notification when a user edits their own comment, and needs to be moderated.
- `notificationModeratorApprovedEnabled` - Whether to notify comment authors when their comment has been approved via moderation.
- `notificationAdmins` - Enter the email address of any administrators, used for the below settings.
- `notificationAdminEnabled` - Admins will receive an email whenever someone makes a comment.
- `notificationFlaggedEnabled` - Users receive an email when someone replies to their comment.
- `useQueueForNotifications` - Whether to use Craft‘s queue system to trigger emails. This is highly recommended to prevent slow comment submission for your users.

- `showCustomFieldNames` - Whether custom fields should show their field names as labels.
- `showCustomFieldInstructions` - Whether custom fields should show their instruction text underneath labels.
- `sortDefaultKey` and `sortDefaultDirection` - Changes the default sort order in the control panel when viewing all comments. (Does not affect the front-end).

## Control Panel
You can also manage configuration settings through the Control Panel by visiting Settings → Comments.

### `notificationAdmins`
Provide a nested array of emails, each with an `enabled` item.

```php
'notificationAdmins' => [
    [
        'email' => 'admin@site.com',
        'enabled' => true,
    ],
    [
        'email' => 'admin-alt@site.com',
        'enabled' => true,
    ],
],
```

### `permissions`
Provide a nested array of element permissions. You'll need to use the UID for your "groups" of elements - Sections for Entries, Category Groups for Categories, etc.

```php
'permissions' => [
    'craft\elements\Asset' => [
        // Volume UIDs
        '33974e79-b3b6-47ec-af91-4519fe4985be',
        'c194716f-aa74-40b6-8426-c835599cbe93',
    ],

    'craft\elements\Category' => [
        // Category Group UIDs
        'a27827c8-4810-433c-acb9-261b53d46281',
        '346194f0-6da8-4f46-a20b-795631ee9a5f',
    ],

    'craft\elements\Entry' => [
        // Section UIDs
        'a8d4bdf1-164e-4ddc-aaad-640026b8d3bf',
    ],

    'craft\elements\User' => [
        // User Group UIDs
        'c23a4d8d-47f0-4e71-927c-d5897ec9c9f8',
    ],
],
```

CHANGELOG.md
# Changelog

## 2.0.9 - 2023-12-03

### Added
- Add action for Comments to determine state of saving or deleting.
- Add `inline` option for reCAPTCHA JS loading.

### Fixed
- Fix an error where trashed comments weren’t validated correctly.
- Fix custom email templates not working for notifications.
- Fix an error with `ownerSiteId` for multi-site installs.
- Fix lack of typecast for `siteId` when submitting comments.

## 2.0.8 - 2023-10-25

### Added
- Add support for Votes to Feed Me itegration. (thanks @JeroenOnstuimig).
- Add extra error-handling for some email notifications.

### Fixed
- Implement `Element::trackChanges()` for Blitz compatibility.
- Fix returning Vote user when both querying users and comments is allowed for GraphQL. (thanks @JeroenOnstuimig).

## 2.0.7 - 2023-03-09

### Changed
- Update system message headings for some email notifications.

### Fixed
- Fix an issue when installing the plugin, and automatically disabled.
- Fix an incompatibility with [Emails](https://github.com/ryssbowh/craft-emails) plugin.
- Fix incorrect return types for voting and flagging comments with GraphQL mutations.

## 2.0.6 - 2023-02-22

### Added
- Add `name` and `email` to available index columns for the control panel.

### Changed
- Update English translations.
- Only admins are now allowed to access plugin settings.
- Improve dutch translations. (thanks @jeroenlammerts).

### Fixed
- Fix a 404 error when trying to edit a non-primary site comment for a multi-site install.
- Fix HTML encoding for name, email and comment columns in the control panel element index.
- Fix compatibility with Craft 4 and guest names showing `User`.
- Fix an error when importing nested commerce with Feed Me.
- Fix email notifications not sending via the queue when comments are made from a non-primary site
- Fix custom sources not filtering by `$ownerSectionId`. (thanks @jeroenlammerts).

## 2.0.5 - 2022-12-30

### Fixed
- Fix an error when sending email notifications.

## 2.0.4 - 2022-12-25

### Changed
- Improve dutch translations. (thanks @jeroenlammerts).

### Fixed
- Fixed PHP errors that could occur when executing GraphQL queries.

## 2.0.3 - 2022-10-12

### Fixed
- Fix an error when upvoting immediately after downvoting.
- Fix avatars not working for logged-in users.

## 2.0.2.1 - 2022-09-17

### Fixed
- Fix a fatal PHP error introduced in `2.0.2`.

## 2.0.2 - 2022-09-17

### Added
- Add support for setting ReCAPTCHA API keys via .env variables.

### Fixed
- Fix template path not resetting when calling `commentsInclude()`.
- Fix an error when trying to edit a comment from the element index.
- Fix element index arrow for comment elements.
- Fix element action for setting status.
- Fix lack of permission support for managing other user’s comments.
- Fix being unable to set `id`, `uid` or `userId` for GraphQL mutations.
- Fix an error running `resave` console commands.

## 2.0.1 - 2022-08-27

### Added
- Added missing English Translations.

### Fixed
- Fix an error when viewing comments with an avatar image.
- Fix formatting for comments element index.
- Fix lack of proper error when trying to edit an invalid comment in the control panel.
- Fix the CP edit URL for comments to now no longer include the site ID.
- Fix an error when importing comments via Feed Me in some instances.
- Fix an error when uninstalling.
- Fix being unable to view comments in the control panel.

## 2.0.0 - 2022-06-02

### Added
- Add resave console command for elements.
- Add checks for registering events for performance.
- Add `archiveTableIfExists()` to install migration.

### Changed
- Now requires PHP `8.0.2+`.
- Now requires Craft `4.0.0+`.
- Now requires Comments `1.9.2` in order to update from Craft 3.
- Rename service classes.
- Rename base plugin methods.

### Fixed
- Fix an error with GraphQL.
- Fix typing for `placeholderAvatar` plugin setting.
- Un-bold title for comments in comments index.
- Fix an error when permissions exist for a no-longer-available element group.
- Fix setting a parent comment not working for Feed Me, due to `setParent()` changes.

### Removed
- Removed `craft.comments.all()` and `craft.comments.form()`.
- Removed `Comment::trashUrl`, `Comment::flagUrl`, `Comment::downvoteUrl`, `Comment::upvoteUrl`.

## 1.9.6 - 2023-03-09

### Changed
- Update system message headings for some email notifications.

### Fixed
- Fix an issue when installing the plugin, and automatically disabled.
- Fix an incompatibility with [Emails](https://github.com/ryssbowh/craft-emails) plugin.

## 1.9.5 - 2023-02-22

### Changed
- Improve dutch translations. (thanks @jeroenlammerts).

### Fixed
- Fix email notifications not sending via the queue when comments are made from a non-primary site
- Fix custom sources not filtering by `$ownerSectionId`. (thanks @jeroenlammerts).

## 1.9.4 - 2022-09-17

### Added
- Add support for setting ReCAPTCHA API keys via .env variables.

### Fixed
- Fix template path not resetting when calling `commentsInclude()`.

## 1.9.3 - 2022-06-02

### Added
- Add `Minimum Score` plugin setting for reCAPTCHA.

### Fixed
- Fix JS render variables being way too large.
- Fix an error saving a comment when the current site can’t be determined.
- Fix replies not showing in the correct order (latest first).

## 1.9.2 - 2022-02-27

### Added
- Add “Use Queue for Notifications” plugin setting (disabled by default).

### Fixed
- Fix SQL errors when ordering comments via "Votes" or "Flagged" in the control panel.
- Fix an error when sorting comments by their "flagged" column in the control panel.
- Fix an error when including "flagged" as a column when viewing comments in the control panel.

## 1.9.1 - 2021-12-31

### Fixed
- Fix Feed Me importing not taking into account trashed owners.

## 1.9.0 - 2021-12-04

### Changed
- Now requires Craft 3.7.24+.

### Fixed
- Fix an incompatibility error with Craft 3.7.24+.

## 1.8.12 - 2021-12-04

### Added

src/models/Settings.php
<?php
namespace verbb\comments\models;

use verbb\comments\Comments;
use verbb\comments\elements\Comment;
use verbb\comments\enums\CommentStatus;

use Craft;
use craft\base\ElementInterface;
use craft\base\Model;
use craft\db\Table;
use craft\elements\Asset;
use craft\elements\User;
use craft\helpers\App;
use craft\helpers\ArrayHelper;
use craft\helpers\Db;
use craft\helpers\Json;

class Settings extends Model
{
    // Properties
    // =========================================================================

    public string $pluginName = 'Comments';
    public ?string $structureUid = null;
    public bool $closed = false;
    public int $indexSidebarLimit = 25;
    public bool $indexSidebarGroup = true;
    public bool $indexSidebarIndividualElements = false;
    public array $defaultQueryStatus = [Comment::STATUS_APPROVED];

    // General
    public bool $allowGuest = false;
    public ?string $guestNotice = null;
    public bool $guestRequireEmailName = true;
    public bool $guestShowEmailName = true;
    public bool $requireModeration = true;
    public ?string $moderatorUserGroup = null;
    public bool $moderatorExcluded = true;
    public mixed $autoCloseDays = null;
    public mixed $maxReplyDepth = null;
    public mixed $maxUserComments = null;

    // Voting
    public bool $allowVoting = true;
    public bool $allowGuestVoting = false;
    public int $downvoteCommentLimit = 5;
    public bool $hideVotingForThreshold = false;

    // Flagging
    public bool $allowFlagging = true;
    public bool $allowGuestFlagging = false;
    public int $flaggedCommentLimit = 5;

    // Templates - Default
    public bool $showAvatar = true;
    public mixed $placeholderAvatar = null;
    public bool $enableGravatar = false;
    public bool $showTimeAgo = true;
    public bool $outputDefaultCss = true;
    public bool $outputDefaultJs = true;

    // Templates - Custom
    public ?string $templateFolderOverride = null;
    public ?string $templateEmail = null;

    // Security
    public bool $enableSpamChecks = true;
    public ?string $securityMaxLength = null;
    public ?string $securityFlooding = null;
    public ?string $securityModeration = null;
    public ?string $securitySpamlist = null;
    public ?string $securityBanned = null;
    public bool $securityMatchExact = false;
    public bool $recaptchaEnabled = false;
    public ?string $recaptchaKey = null;
    public ?string $recaptchaSecret = null;
    public float $recaptchaMinScore = 0.5;

    // Notifications
    public bool $notificationAuthorEnabled = true;
    public bool $notificationReplyEnabled = true;
    public bool $notificationSubscribeAuto = false;
    public bool $notificationSubscribeDefault = true;
    public bool $notificationSubscribeEnabled = false;
    public bool $notificationSubscribeCommentEnabled = false;
    public bool $notificationModeratorEnabled = false;
    public bool $notificationModeratorEditEnabled = false;
    public bool $notificationModeratorApprovedEnabled = false;
    public array $notificationAdmins = [];
    public bool $notificationAdminEnabled = false;
    public bool $notificationFlaggedEnabled = false;
    public bool $useQueueForNotifications = false;

    // Permissions
    public array $permissions = [];

    // Users
    public array $users = [];

    // Custom Fields
    public bool $showCustomFieldNames = false;
    public bool $showCustomFieldInstructions = false;

    // CP Sorting
    public string $sortDefaultKey = 'structure';
    public string $sortDefaultDirection = 'asc';

    private ?Asset $_placeholderAvatar = null;


    // Public Methods
    // =========================================================================

    public function getPlaceholderAvatar(): ?Asset
    {
        if ($this->_placeholderAvatar !== null) {
            return $this->_placeholderAvatar;
        }

        if ($this->placeholderAvatar && isset($this->placeholderAvatar[0])) {
            /* @noinspection PhpIncompatibleReturnTypeInspection */
            return $this->_placeholderAvatar = Craft::$app->getElements()->getElementById($this->placeholderAvatar[0], Asset::class);
        }

        return null;
    }

    public function canComment($element): bool
    {
        $isAllowed = $this->commentingAvailable($element);

        return $isAllowed['permission'];
    }

    public function commentingAvailable($element): array
    {
        $isClosed = Comments::$plugin->getComments()->checkManuallyClosed($element);

        if ($isClosed) {
            return CommentStatus::ManuallyClosed;
        }

        $isExpired = Comments::$plugin->getComments()->checkExpired($element);

        if ($isExpired) {
            return CommentStatus::Expired;
        }

        $hasPermission = Comments::$plugin->getComments()->checkPermissions($element);

        if (!$hasPermission) {
            return CommentStatus::Unpermitted;
        }

        $currentUser = Comments::$plugin->getService()->getUser();

        if (!$currentUser && !$this->allowGuest) {
            return CommentStatus::NoGuests;
        }

        if ($this->maxUserComments && $currentUser) {
            // Has the user already commented X amount of times on this element?
            $count = Comment::find()->ownerId($element->id)->userId($currentUser->id)->count();

            if ($count >= $this->maxUserComments) {
                return CommentStatus::TooManyComments;
            }
        }

        return CommentStatus::Allowed;
    }

    public function getStructureId(): ?int
    {
        if ($this->structureUid) {
            return Db::idByUid(Table::STRUCTURES, $this->structureUid);
        }

        // Create the structure if it doesn't exist
        if ($structure = Comments::$plugin->createAndStoreStructure()) {
            $this->structureUid = $structure->uid;

            return $structure->id;
        }

        return null;
    }

    public function getEnabledNotificationAdmins(): array
    {
        $notificationAdmins = $this->notificationAdmins ?: [];

        return ArrayHelper::where($notificationAdmins, 'enabled');
    }

    public function getRecaptchaKey()
    {
        return App::parseEnv($this->recaptchaKey);
    }

    public function getRecaptchaSecret()
    {
        return App::parseEnv($this->recaptchaSecret);
    }

    public function doesRequireModeration(): bool
    {
        // Check if we require moderation at all
        if ($this->requireModeration) {
            // Check if we should exclude the current user, if they are in the moderator group
            if ($this->moderatorExcluded && $this->moderatorUserGroup) {
                if ($currentUser = Comments::$plugin->getService()->getUser()) {
                    $groupId = Db::idByUid(Table::USERGROUPS, $this->moderatorUserGroup);
                    $moderators = User::find()->groupId($groupId)->ids();

                    if (in_array($currentUser->id, $moderators)) {
                        return false;
                    }
                }
            }

            return true;
        }

        return false;
    }
}
